### ISSUE TITLE
this.router.navigate(...) doesn't work in IE9

### ISSUE THREAD

**Issue by coreyog** [angular/angular#8160]
**Steps to reproduce and a minimal demo of the problem**

Open this plnkr in IE9.

http://plnkr.co/edit/IJ6tcuVOb5npbElvQzbq?p=preview

`this.router.navigate(...);` in home.component.ts does nothing at all when you click the button. It should tack on a #/test route to the URL and show the other component. 

_What steps should we try in your demo to see the problem?_

Using IE9, click the only button on the screen to take you to the Test view. You'll notice nothing shows up in the console and the routing doesn't happen. The page content doesn't change and the URL doesn't update to show the new route. In Chrome or later versions of IE it will successfully route to the new view.

**Expected/desired behavior**

Routing should update URL and page contents in IE9.

**Other information**

There is zero information in the console. No indication at all for a failure.

**Comment by ericmartinezr** [comment]
I'm getting `pushState` error in the console with your plnkr. I removed your scripts and let the ones you commented so I have the official order. Regarding the error, you may want to see this issue https://github.com/angular/angular/issues/6506

**Comment by coreyog** [comment]
I am not getting the pushState error nor any error on my test machine. Maybe it's a different minor version of IE? I'm using 9.0.8112.16421.

The shims I have in the order I have them is what works for me. I don't know why the official ones are giving me such an issue.

**Comment by ericmartinezr** [comment]
I'm getting this error 

![image](https://cloud.githubusercontent.com/assets/13039652/14690691/ac833ec0-0723-11e6-9098-a8e071b9e47f.png)

Translation :
*Object doesn't support property or method 'pushState'*

Note that I'm using  IE11 with IE9 emulated. I tried your plnkr again and I saw that you removed the official order, and I still get the error, so it's not an order issue as far I can tell.

Have you tried the plnkr in another machine?

**Comment by coreyog** [comment]
That's the issue. I'm using an actual IE 9 and I get no errors:

![image](https://cloud.githubusercontent.com/assets/2836207/14690836/66e6bed6-070b-11e6-92f5-d80c230989f6.png)

**Comment by ericmartinezr** [comment]
I downloaded Win 7 + IE9 from this [link](https://www.microsoft.com/en-us/download/details.aspx?id=11575) and using a real IE9 I get the error with your plnkr. See the [screenshot](https://cloud.githubusercontent.com/assets/13039652/14692370/20e7db64-072d-11e6-8676-291491a338f2.png) and IE version is in this [screenshot](https://cloud.githubusercontent.com/assets/13039652/14692397/46dbe0cc-072d-11e6-8272-4cf3fc628084.png). At this point I don't know what else could be =/.

**Comment by sod** [comment]
http://caniuse.com/#search=pushState says IE9 doesn't support pushState. My current solution is sham'ing pushState, if it's missing

```typescript
const _history: History = window.history;
const _location: Location = window.location;

// IE9 pushState fallback
if (!_history.pushState) {
   _history.pushState = _history.replaceState = function(state: any, title: string, url: string) {
        if (url !== _location.pathname) {
            _location.pathname = url;
        }
    };
}
```

**Comment by zoechi** [comment]
proposed fix #7929 (dup of #6506 as mentioned above)

**Comment by coreyog** [comment]
@sod the shim worked for me.

I'm confused why I'm not getting anything in the console about pushState not being in IE9 but oh well. It's all working as expected now. I'll just add this to my ie.js:

```
(function() {
  var _history = window.history;
  var _location = window.location;

  // IE9 pushState fallback
  if (!_history.pushState) {
    _history.pushState = _history.replaceState = function(state, title, url) {
          if (url !== _location.pathname) {
              _location.pathname = url;
          }
      };
  }
}());
```

It's just converted to JS and wrapped to be self executing. Maybe it'll save somebody else some headaches.
